src/v/cluster/rm_stm.cc

```diff
@@ -1310,15 +1306,9 @@ model::offset rm_stm::last_stable_offset() {
         if (!_log_state.ongoing_set.empty()) {
             first_tx_start = *_log_state.ongoing_set.begin();
         }
-
-        for (auto& entry : _mem_state.estimated) {
-            first_tx_start = std::min(first_tx_start, entry.second);
-        }
```

Comment:
It looks that removing this and keeping the current logic isn't safe. Consider a following situation:

- replicate fence batch with offset 10 (begining tx)
- replicate some tx batches that belongs to started tx
If the fence batch wasn't yet applied and caller queries for `last_stable_offset` it may be the case we would return an `lso = model::next_offset(last_visible_index);` as there is no information about ongoing transaction available.

I totally agree with the sentiment of removing the in memory state however i think we should still leverage the advantage that the `raft::persisted_stm` gives use. With the persisted STM after we call `sync()` we are sure that current node is the leader and it is up to date with all information in the log which means that all the batches were applied to the STM. After this we may leverage the fact that the leader is the only one making the state changes and update part of the state before the batches are applied to the STM or validate some state before we even call Replicate. We need to remember however that we would have to revert the changes if replicate call fails or `sync()` fails on the leader.

Here in `last_stable_offset()` we are not calling sync() explicitly so we can not rely on the fact that all data were applied. I think it is fine to simply ignore the last visible index here and only rely on what was already applied to the STM which would make this code safe.